File: task3/C/task3.cpp
/* На вход подаётся N натуральных чисел, для каждого трёхзначного
числа определить, является ли оно палиндромом (читается одинаково с любой
стороны).
Пример 1: N = 4, 101 456 707 213, Результат: Yes No Yes No
Пример 2: N = 3, 1221 676 33, Результат: - Yes - */

#include <iostream>
#include <vector>

using namespace :: std;

bool isPalindrome(int number) { 
    int cpnum = number;
    int reverseNumber = 0;
    while (number){
        reverseNumber = reverseNumber * 10 + number % 10;
        number /= 10;
    }
     return reverseNumber == cpnum;
}

vector<string> palindromesInArray(vector<int> numbers) {
    vector<string> result;
    for (size_t i = 0; i < numbers.size(); ++i) {
        if (isPalindrome(numbers[i]) && numbers[i] < 1000 && numbers[i] > 99) result.push_back("Yes");
        else if (!isPalindrome(numbers[i]) && numbers[i] < 1000 && numbers[i] > 99) result.push_back("No");
        else result.push_back("-");
    }
    return result;
}

template <typename T>
void display(vector<T> v) { //функция вывода вектора
    for (size_t i = 0; i < v.size(); ++i) {
        cout << v[i] << " ";
    }
}

int main() {
    int N;
    cout << "Enter N: ";
    cin >> N;
    if (N < 0) {
        cout << "N should be positive!!!" << endl;
        return 0;
    }
    vector<int> numbers(N);
    cout << "Enter numbers: " << endl;
    for (size_t i = 0; i < N; ++i) {
        cout << (i + 1) << " - ";
        cin >> numbers[i];
        if (numbers[i] < 0){
            cout << "Number should be positive!!!" << endl;
            return 0;
        }
    }
    display(palindromesInArray(numbers));
    return 0;
}

File: task3/Elixir/task3.exs
defmodule Palindrome do
  def is_palindrome(number) do
    reversed =
      number
      |> Integer.digits()
      |> Enum.reverse()
      |> Integer.undigits()

    reversed == number
  end

  def palindromes_in_array(numbers) do
    Enum.map(numbers, fn num ->
      cond do
        num > 99 and num < 1000 ->
          if is_palindrome(num), do: "Yes", else: "No"

        true ->
          "-"
      end
    end)
  end

  def display(list) do
    Enum.each(list, fn item ->
      IO.write("#{item} ")
    end)
  end
end

defmodule Main do
  def run do
    n =
      IO.gets("Enter N: ")
      |> String.trim()
      |> String.to_integer()

    if n < 0 do
      IO.puts("N should be positive!!!")
    else
      numbers =
        Enum.map(1..n, fn i ->
          input =
            IO.gets("#{i} - ")
            |> String.trim()
            |> String.to_integer()

          if input < 0 do
            IO.puts("Number should be positive!!!")
            System.halt(0)
          end

          input
        end)

      numbers
      |> Palindrome.palindromes_in_array()
      |> Palindrome.display()
    end
  end
end

Main.run()
